skimage/morphology/tests/test_flood_fill.py

```diff
@@ -0,0 +1,197 @@
+import numpy as np
+from pytest import raises
+
+from skimage.morphology import flood, flood_fill
+
+eps = 1e-12
+
+
+def test_empty_input():
+    # Test shortcut
+    output = flood_fill(np.empty(0), (), 2)
+    assert output.size == 0
+
+    # Boolean output type
+    assert flood(np.empty(0), ()).dtype == np.bool
+
+    # Maintain shape, even with zero size present
+    assert flood(np.empty((20, 0, 4)), ()).shape == (20, 0, 4)
+
+
+def test_float16():
+    image = np.array([9., 0.1, 42], dtype=np.float16)
+    with raises(TypeError, match="dtype of `image` is float16"):
+        flood_fill(image, 0, 1)
+
+
+def test_overrange_tolerance_int():
+    image = np.arange(256, dtype=np.uint8).reshape((8, 8, 4))
+    expected = np.zeros_like(image)
+
+    output = flood_fill(image, (7, 7, 3), 0, tolerance=379)
+
+    np.testing.assert_equal(output, expected)
```

Comment:
`np.testing` should not be used directly. We have all the necessary wrappers in https://github.com/scikit-image/scikit-image/blob/master/skimage/_shared/testing.py .

Reply:
I thought that we'd reversed this policy? I remember the comment that having our own testing functions is one more thing to remember "just for skimage", and it resonated very strongly with me. We should use standard things except where it is an undue burden. I realise that the nose -> pytest transition was traumatic, but imho that is not sufficient justification to do our own thing, as we don't expect those transitions to be frequent.

Reply:
(Apologies if the above comment comes like 18 months too late or so. =P)

Reply:
Why will anyone want to revert that? 😲 Frankly speaking, I'm hearing about this for the first time.

To me, the thin wrappers in `skimage._shared.testing` aren't just for potential convenient switch of the testing library. Besides that:

1. It is a single place where all the utilised testing functions and classes are presented. Even after our transition to `pytest` we kept using a lot of functionality from `numpy.testing`. This is due to the fact that neither of those libraries include all the necessary functions and one is not superset of the other. On top of this, `numpy.testing` was still using both `nose` and `pytest` under the hood, and for a skimage contributor it was a pain to understand which functions he or she can import from numpy. Having this thin wrappers, a contributor can just import `skimage._shared.testing` and enjoy the coding; 
2. Single point of deprecation. When `pytest` or `numpy.testing` will have some functionality deprecated, for us it is much easier to rewrite couple of lines in `skimage._shared.testing` than to change few hundreds of them all over the library;
3. Almost-0 maintenance cost.

If you have a strong argument why we should drop that, I'm happy to discuss, but, otherwise, I'd strongly prefer to keep things as is.

Reply:
@soupault I think I just got confused by the discussion [here](https://github.com/scikit-image/scikit-image/pull/3164), specifically [Mark's comment](
https://github.com/scikit-image/scikit-image/pull/3164#issuecomment-395649448). But, my overall feeling remains, so I'll address your points:

1. I get this, I just don't think that scikit-image should be that one place. `np.testing` is a much more logical place for that one place, especially since `_shared` is private, which means we can change things when we feel like it, which means that no outside project can use our wrappers.
2. In contrast, `np.testing` is a public API that is much less likely to be deprecated.
3. I would argue that postponing this PR (and countless others because everyone in the SciPy ecosystem is trained to use `np.testing` and `with pytest.raises`) is a maintenance cost.

Re (3), note that we don't have any reference to `skimage._shared.testing` in the contributing guide.

Reply:
Agree with @jni.

I think I was deeply mired in surgical intern year when the conversion to pytest happened.  These shared wrappers were unknown to me.  Looking around the package, our usage remains heterogeneous.  I would have been :-1: at converting anything to using them, unless they bring new important functionality, and regret not being part of that discussion.

I strongly dislike forcing use of a hidden test API when NumPy's is perfectly suitable.  Where certain features may not exist at base in NumPy, we should publicly expose these (even if they are deeply nested, like `skimage.util.testing` or similar.  The potential portability advantages are not significant enough to justify this.  This is the same reason we generally don't reimplement `scipy.ndimage`.

Imagine a new user making a first contribution.  They generate a perfectly compliant test suite using guides they find online, via NumPy's suite.  We tell them - despite technical correctness - that we won't accept it?  Yikes.  All this really does is provide additional barriers to contribution.

Barriers to contribution are high enough already, IMHO, and we should be looking to reduce them.

Reply:
Background here: we switched before NumPy, so we had to take care of situations where they used nose specific functionality.  The idea behind the wrappers were that there would be a single place to modify things, instead of using pytest-specific decorators, etc.

But, yes, I don't see any reason why we need to maintain our own testing utilities, if the functionality overlaps almost identically with numpy.